src/Core/Traits/StatefulData.php

```diff
@@ -183,14 +184,42 @@ public function setState(array $data)
 				continue;
 			}
 
-			if ($this->$key !== $value) {
+			if (is_array($value)){
+				$current_key = is_array($this->$key) ? $this->$key : [];
+
+				// Check for multi level recursion
+				$diff = array_merge(
```

Comment:
I wonder if its worth hunting for a more efficient method of recursive diff'ing before this lands? ie. a diff thats already notes diffs from both sides, rather than running it twice